aten/src/ATen/native/DilatedConvolution.cpp

```diff
@@ -0,0 +1,879 @@
+        grad_input_n = grad_input.select(0, elt);
+        col2vol<scalar_t>(
+            columns.data<scalar_t>(),
+            nInputPlane,
+            inputDepth,
+            inputHeight,
+            inputWidth,
+            outputDepth,
+            outputHeight,
+            outputWidth,
+            kD,
+            kH,
+            kW,
+            padD,
+            padH,
+            padW,
+            dD,
+            dH,
+            dW,
+            dilationD,
+            dilationH,
+            dilationW,
+            grad_input_n.data<scalar_t>());
+      }
+
+      // Gradient of weight:
+      if (grad_weight.defined()) {
+        scalar_t scale = 1; // TODO: expose as argument?
+        // Extract columns:
+        vol2col<scalar_t>(
+            input_n.data<scalar_t>(),
+            nInputPlane,
+            inputDepth,
+            inputHeight,
+            inputWidth,
+            outputDepth,
+            outputHeight,
+            outputWidth,
+            kD,
+            kH,
+            kW,
+            padD,
+            padH,
+            padW,
+            dD,
+            dH,
+            dW,
+            dilationD,
+            dilationH,
+            dilationW,
+            columns.data<scalar_t>());
+        // For gemm argument derivation, see
+        // conv_dilated2d_all_cuda_template in
+        // ATen/native/cuda/DilatedConvolution.cu
+        THBlas_gemm<scalar_t>(
+            /*transa=*/'t',
+            /*transb=*/'n',
+            /*     m=*/columns.size(0),
+            /*     n=*/nOutputPlane,
+            /*     k=*/columns.size(1),
+            /* alpha=*/scale,
+            /*     A=*/columns.data<scalar_t>(),
+            /*   lda=*/columns.size(1),
+            /*     B=*/grad_output_n.data<scalar_t>(),
+            /*   ldb=*/columns.size(1),
+            /*  beta=*/1,
+            /*     C=*/grad_weight.data<scalar_t>(),
+            /*   ldc=*/columns.size(0));
+      }
+
+      // Gradient of bias:
+      if (grad_bias.defined()) {
+        grad_bias += grad_output_n.sum({1, 2, 3});
+      }
+    }
+  });
+  if (is_batch) {
+    DROP_BATCH_DIMENSION(input);
+    DROP_BATCH_DIMENSION(output);
+    DROP_BATCH_DIMENSION(grad_input);
+    DROP_BATCH_DIMENSION(grad_output);
+  }
+
+} // conv_dilated3d_all_cpu_template
+
+} // namespace
+
+Tensor conv_dilated2d_cpu(
+    const Tensor& input,
+    const Tensor& weight,
+    IntArrayRef kernel_size,
+    const Tensor& bias,
+    IntArrayRef stride_size,
+    IntArrayRef pad_size,
+    IntArrayRef dilation_size) {
+  auto options = input.options();
+  Tensor output = at::empty({0}, options);
+  const Tensor weight_ = weight.contiguous();
+  const Tensor input_ = input.contiguous();
+  const Tensor bias_ = bias.contiguous();
+  Tensor undefined;
+  int64_t nOutputPlane = weight.size(0);
+  int64_t outputHeight = OUTPUTSIZE(0, 2);
+  int64_t outputWidth = OUTPUTSIZE(1, 2);
+  output.resize_({nOutputPlane, outputHeight, outputWidth});
```

Comment:
nit: If you're going to immediately resize the output, might as well just allocate it with the right shape in the first place.

```
Tensor output = at::empty({nOutputPlane, outputHeight, outputWidth}, options);
```

Reply:
However, there is a broader problem: I think you are resizing `output` too early here. You are relying on the result of `OUTPUTSIZE` computation, before you've checked that it is actually a valid size. If `OUTPUTSIZE` produces a negative number, you will error here, rather than in the shape check (that is later). In general, you shouldn't resize output until you've finished error checking. So just allocate the zero-size one and let the inner template handle resizing.

Reply:
Agreed. Checking of argument shapes and data structure has been revised now.